gobblin-core/src/main/java/gobblin/source/extractor/extract/sftp/SftpFsHelper.java

```diff
@@ -51,10 +55,52 @@ public SftpFsHelper(State state) {
     this.state = state;
   }
 
-  public ChannelSftp getSftpConnection() {
-    return this.channelSftp;
+  /**
+   * Create new channel every time a command needs to be executed. This is required to support execution of multiple
+   * commands in parallel. All created channels are cleaned up when the session is closed.
+   * @see #cleanupSftpChannels()
+   * @see #close()
+   *
+   * @return a new {@link ChannelSftp}
+   * @throws SftpException
+   */
+  public ChannelSftp getSftpChannel() throws SftpException {
```

Comment:
This may be the time to start switching to Java 7. I suggest that we return a wrapper CloseableChannelSftp which implements the AutoCloseable interface. That will allow us to write something like:

try (CloseableChannelSftp channel = helper.getSfpChannel()) {
  channel.foo();
} 

That will take care of the channel management.